### ISSUE TITLE
Deprecate on(Boolean), where(Boolean), having(Boolean), etc.

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#4763]
With predicate-accepting methods accepting `Boolean` for convenience, we might have an issue with `Object.equals()` vs. `Field.equal()` confusion. It's now possible to accidentally write:

```java
join(CUSTOMER).on(CUSTOMER.ID.equals(ORDERS.CUSTOMER_ID)) // empty result
where(CUSTOMER.ID.equals(customerId)) // empty result
having(count().equals(1)) // empty result
```

We might need to remove these methods again (deprecating them first), possibly also in 3.7.2

----
See also:
https://github.com/jOOQ/jOOQ/issues/4366#issuecomment-160741310

**Comment by lukaseder** [comment]
I have thought about this some more, and I think that it is really valueable to have this overload on all these methods. The real issue here is the `equal()` method, which is too similar to `equals()`. I have also thought about deprecating `Field.equals()` in order to produce compiler warnings as people hardly ever want to use that method (except in other equals implementations).

So, we have three options:

- Deprecating (and later removing) these `where(Boolean)` methods again
- Deprecating (and later removing) `Field.equal()`
- Deprecating (for documentation, not removal) `Field.equals()`

// cc @stellingsimon

**Comment by stellingsimon** [comment]
In Eclipse, if I type `FIELD.eq(`, it is auto-expanded to `FIELD.equals(o)`, so as long as the `where(Boolean)` overload exists, I will experience this nuisance when writing ordinary conditions.

Maybe there's a fourth option? 

 * Deprecating (and later removing) these `where(Boolean)` methods again, replacing them with `whereTrue()`, `onTrue()` etc.?

**Comment by lukaseder** [comment]
I understand the problem, and it is really weird that `FIELD.equals()` is given such a high priority by Eclipse. But good feedback, that kind of invalidates Option 2 (which wasn't my favourite anyway)

The fourth option won't do it in my opinion. The idea of `where(Boolean)` is that you can really pass any boolean, including a variable. A `whereTrue()` won't be very useful.

**Comment by lukaseder** [comment]
I've tried to bend my head around this several times now, justifying keeping the `xxx(Boolean)` methods in one way or another. But in the end, I'm always ending up with pondering the added value of the methods against the nuisance this causes to you and others.

jOOQ's success is due to the fact that it is an extremely user friendly API, causing almost no cognitive friction when writing SQL in Java.

Having to remember the difference between `equal` and `equals` and not having the compiler remind you of this difference is cognitive friction - at a rather central aspect of the DSL.

I'll conclude with deprecating these methods again, removing them in jOOQ 4.0. They were a mistake / oversight on our part. Their functionality can be achieved by wrapping boolean values with `DSL.val(boolean)`:

```java
// instead of
.where(true)

// write:
.where(val(true))
```

**Comment by lukaseder** [comment]
The issue and nuisance is significant enough to be merged also to version 3.7.2 (#4816). With the deprecation, at least, when accidentally using `where(Boolean)`, a compiler warning is issued.

**Comment by lukaseder** [comment]
5 years later, these deprecated methods will now be removed in jOOQ 3.15.0: https://github.com/jOOQ/jOOQ/issues/11242
